### PR TITLE
Fix intra-doc links for inherent impls that are both lang items and not the default impl

### PR DESCRIPTION
**Pull request by jyn514**
I found in https://github.com/rust-lang/rust/pull/75464#issuecomment-675125984 that `str::to_uppercase()` doesn't resolve while `str::trim()` does. The only real difference is that `to_uppercase` is defined in `alloc`, while trim is defined in `core`. It turns out that rustdoc was ignoring `lang_items.str_alloc_impl()` - it saw them in `collect_trait_impls`, but not for intra-doc links.

This uses the same `impls` for all parts of rustdoc, so that there can be no more inconsistency. It does have the slight downside that the matches are no longer exhaustive but it will be very clear if a new lang item is missed because it will panic when you try to document it (and if you don't document it, does rustdoc really need to know about it?).

~~This needs a test case (probably just `str::to_uppercase`).~~ Added.

This is best reviewed commit-by-commit.

r? @GuillaumeGomez

### CODE REVIEW

**Comment by jyn514** [comment]
Oops, this also has the changes from #74489. I need to remove those, they're independent.

**Comment by jyn514** [comment]
This does keep one small change from #74489, which is to use `find_by_name_and_namespace` instead of `filter_by_name_unhygenic`. But that's a good change anyway and meant fewer rebase conflicts.

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -1264,6 +1266,61 @@ impl PrimitiveType {
...
+            }
+            Reference => None,
+            Fn => None,
+            Never => None,
+        };
+
+        primary_impl.into_iter().collect()
+    }
+
+
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/types.rs]: Please remove this empty line.

**Comment by GuillaumeGomez** [comment]
Nice work!

@bors: r+

**Comment by tmandry** [comment]
Failed in https://github.com/rust-lang/rust/pull/75688#issuecomment-675853876 due to Cargo.lock being out of date
@bors r-

**Comment by jyn514** [comment]
@bors r-

I want to add a test for f32 and f64 methods.

**Comment by jyn514** [comment]
@bors r=@guillaumegomez

Tagging you so you can look at the changes if you want :) but it's only adding a new test.

**Pull request merged**